keepsorted/line_group.go

```diff
@@ -15,25 +15,95 @@
 package keepsorted
 
 import (
-	"cmp"
 	"fmt"
 	"regexp"
 	"strings"
+	"sync"
 	"unicode"
 
 	"github.com/rs/zerolog/log"
 )
 
-// lineGroup is a logical unit of source code. It's one or more lines combines
+// lineGroup is a logical unit of source code. It's one or more lines combined
 // with zero or more comment lines about the source code lines.
 type lineGroup struct {
+	opts        blockOptions
+	prefixOrder func() *prefixOrder
+
+	// The actual content of the lineGroup.
+	lineGroupContent
+
+	// Bits of information that we use while comparing two line groups together.
+	// These bits of information are calculated lazily, and are invalidated when
+	// the lineGroup is mutated. You should prefer to access them via their
+	// getters.
+	calculated lineGroupCalculations
+}
+
+var compareLineGroups = comparingFunc((*lineGroup).commentOnly, falseFirst()).
+	andThen(comparingFunc((*lineGroup).regexTokens, lexicographically(compareRegexTokens))).
+	andThen(comparing((*lineGroup).joinedLines)).
+	andThen(comparing((*lineGroup).joinedComment))
+
+var compareRegexTokens = comparingFunc(func(t regexToken) bool { return t == nil }, falseFirst()).
+	andThen(comparingFunc(func(t regexToken) []*captureGroupToken { return t }, lexicographically(compareCaptureGroupTokens)))
+
+var compareCaptureGroupTokens = comparingFunc((*captureGroupToken).prefix, orderedPrefix.compare).
+	andThen(comparingFunc((*captureGroupToken).transform, numericTokens.compare))
+
+type lineGroupContent struct {
 	comment []string
 	lines   []string
 }
 
+type lineGroupCalculations struct {
+	commentOnly *bool
+	regexes     []regexToken
+	lines       string
```

Comment:
When this rolls out, should probably keep an eye on resource usage. I expect this should not increase memory usage since we had to come up with this anyway, but something to watch out for.